runtime/src/utxo.rs

```diff
@@ -144,199 +100,152 @@ decl_module! {
-            } else {
-                missing_utxo.push(&input.parent_output);
-            }
+            let utxo = <UtxoStore>::get(&input.outpoint).ok_or("missing input utxo")?;
+
+            ensure!(sp_io::crypto::sr25519_verify(
+                        &Signature::from_raw(*input.sigscript.as_fixed_bytes()),
+                        &simple_transaction,
+                        &Public::from_h256(utxo.pubkey)
+                    ), "signature must be valid"
+            );
+            
+            total_input = total_input.checked_add(utxo.value).ok_or("input value overflow")?;
         }
 
-        let mut total_output: Value = 0;
+        let index: u64 = 0;
         for output in _transaction.outputs.iter() {
             ensure!(output.value != 0, "output value must be nonzero");
+            
+            let hash = BlakeTwo256::hash_of(&(&_transaction.encode(), index));// BlakeTwo256::hash_of(output);
+            index.checked_add(1).ok_or("output index overflow")?;
+            ensure!(!<UtxoStore>::exists(hash), "output already exists");
+            
+            total_output = total_output.checked_add(output.value).ok_or("output value overflow")?;
+        }
 
-            let hash = BlakeTwo256::hash_of(output);
-            ensure!(!<UnspentOutputs>::exists(hash), "output already exists");
+        ensure!( total_input >= total_output, "output value must not exceed input value");
 
-            total_output = total_output
-                .checked_add(output.value)
-                .ok_or("output value overflow")?;
+        Ok( total_input.checked_sub(total_output).ok_or("reward underflow")? )
+    }
+    
+    /// Update storage to reflect changes made by transaction
+    /// Where each utxo key is a hash of the entire transaction and its order in the TransactionOutputs vector
+    fn update_storage(transaction: &Transaction, reward: Value) -> DispatchResult {
+        // Calculate new reward total
+        let new_total = <RewardTotal>::get()
+            .checked_add(reward)
+            .ok_or("Reward overflow")?;
+        <RewardTotal>::put(new_total);
+
+        // Storing updated reward value
+        for input in &transaction.inputs {
+            <UtxoStore>::remove(input.outpoint);
         }
 
-        if missing_utxo.is_empty() {
-            ensure!(
-                total_input >= total_output,
-                "output value must not exceed input value"
-            );
-            Ok(CheckInfo::Totals {
-                input: total_input,
-                output: total_output,
-            })
-        } else {
-            Ok(CheckInfo::MissingInputs(missing_utxo))
+        // Add new UTXO to be used by future transactions
+        let index: u64 = 0;
+        for output in &transaction.outputs {
+            let hash = BlakeTwo256::hash_of(&(&transaction.encode(), index));
+            <UtxoStore>::insert(hash, output);
+            index.checked_add(1).ok_or("output index overflow")?;
```

Comment:
good pt, i moved up the checked_add to before write to storage

its just a style thing, i prefer to keep in all the check_add/subs instead of doing a ++ or -- in any substrate code.